sourceaddrs/source_final_test.go

```diff
@@ -106,3 +106,70 @@ func TestResolveRelativeFinalSource(t *testing.T) {
 		})
 	}
 }
+
+func TestParseFinalSource(t *testing.T) {
+	onePointOh := versions.MustParseVersion("1.0.0")
+
+	tests := []struct {
+		Addr    string
+		Want    FinalSource
+		WantErr string
+	}{
+		{
+			Addr: "./a/b",
+			Want: MustParseSource("./a/b").(FinalSource),
+		},
+		{
+			Addr: "git::https://github.com/hashicorp/go-slug.git//beep/boop",
+			Want: MustParseSource("git::https://github.com/hashicorp/go-slug.git//beep/boop").(FinalSource),
+		},
+		{
+			Addr: "example.com/foo/bar/baz@1.0.0//beep",
+			Want: MustParseSource("example.com/foo/bar/baz//beep").(RegistrySource).Versioned(onePointOh),
+		},
+		{
+			Addr:    " ./a/b",
+			WantErr: "source address must not have leading or trailing spaces",
+		},
+		{
+			Addr:    "",
+			WantErr: "a valid source address is required",
+		},
+		{
+			Addr:    "example.com/foo/bar/baz@1.0.x//beep",
+			WantErr: `invalid module registry source address "example.com/foo/bar/baz@1.0.x//beep": invalid version: can't use wildcard for patch number; an exact version is required`,
+		},
+	}
```

Comment:
I suggest also some local and remote addresses that have at signs in their path portions, to make sure they don't get misclassified as being registry addresses.

For example, `./foo/bar/baz@1.0.0` should presumably match as a local source whose final path segment happens to contain an at sign. I expect that test will already pass but I'd like to ensure it doesn't get broken in future maintenance, since this sort of thing was already a maintenance hazard for the similar code in Terraform Core's `addrs` package.

Reply:
Done!